File: leetcode/160-intersection-of-two-linked-lists.js
// https://leetcode.com/problems/intersection-of-two-linked-lists/ (top-interview-questions)

// Definition for singly-linked list.
function ListNode(val) {
  this.val = val;
  this.next = null;
}

/**
 * @param {ListNode} headA
 * @param {ListNode} headB
 * @return {ListNode}
 */
var getIntersectionNode = function(headA, headB) {

  let tail = headA;
  while (tail.next) {
    tail = tail.next;
  }

  tail.next = headB;

  // cycle detection algorithm...
  let tortoise = headA;
  let hare = headA;

  while (hare && hare.next) {
    tortoise = tortoise.next;
    hare = hare.next?.next;
    if (tortoise === hare) break; 
  }

  // check if there is no cycle at all
  if (hare === null || hare.next === null) {
    tail.next = null;
    return null;
  }

  let tortoise2 = headA;

  while (tortoise != tortoise2) {
    tortoise = tortoise.next;
    tortoise2 = tortoise2.next;
  }

  tail.next = null;
  return tortoise;

};

File: leetcode/160-intersection-of-two-linked-lists.py
# https://leetcode.com/problems/intersection-of-two-linked-lists/ (top-interview-questions)

from typing import Optional


# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    def getIntersectionNode(self, headA: ListNode, headB: ListNode) -> Optional[ListNode]:
        tail = headA
        while tail.next: tail = tail.next
        tail.next = headA
        slow, fast = headB, headB
        while fast and fast.next:
            fast = fast.next.next
            slow = slow.next
            if slow is fast: break

        if fast is None or fast.next is None:
            tail.next = None
            return None

        slow2 = headB
        while slow != slow2:
            slow = slow.next
            slow2 = slow2.next
        tail.next = None
        return slow
